bitswap/provider.go

```diff
@@ -176,61 +176,41 @@ func (provider *Provider) ReceiveMessage(ctx context.Context, sender peer.ID, in
 		// If retriever is disabled, nothing we can do, send DONT_HAVE and move
 		// on
 		if provider.retriever == nil {
-			provider.taskQueue.PushTasks(sender, peertask.Task{
-				Topic:    topicDontHave(entry.Cid),
-				Priority: int(entry.Priority),
-				Work:     message.BlockPresenceSize(entry.Cid),
-			})
+			provider.queueDontHave(ctx, sender, entry, "disabled_retriever")
 			continue
 		}
 
 		// At this point, the blockstore did not have the requested block, so a
 		// retrieval is attempted
+
+		// Record that a retrieval is required
 		stats.Record(ctx, metrics.BitswapRetrieverRequestCount.M(1))
+
 		switch entry.WantType {
 		case wantTypeHave:
 			// TODO: for WANT_HAVE, just check if there's a candidate, we
 			// probably don't have to actually do the retrieval yet
-
 			if err := provider.retriever.Request(entry.Cid); err != nil {
 				// If no candidates were found, there's nothing that can be done, so
 				// queue DONT_HAVE and move on
-				provider.taskQueue.PushTasks(sender, peertask.Task{
-					Topic:    topicDontHave(entry.Cid),
-					Priority: int(entry.Priority),
-					Work:     message.BlockPresenceSize(entry.Cid),
-				})
+				provider.queueDontHave(ctx, sender, entry, "failed_retriever_request")
 				continue
 			}
 
 			provider.blockManager.GetAwait(ctx, entry.Cid, func(block blocks.Block) {
-				stats.Record(context.Background(), metrics.BitswapResponseCount.M(1))
-				provider.taskQueue.PushTasks(sender, peertask.Task{
-					Topic:    topicHave(block.Cid()),
-					Priority: int(entry.Priority),
-					Work:     message.BlockPresenceSize(entry.Cid),
-				})
+				provider.queueHave(ctx, sender, entry, block.Cid())
 				provider.signalWork()
 			})
 		case wantTypeBlock:
 			if err := provider.retriever.Request(entry.Cid); err != nil {
 				// If no candidates were found, there's nothing that can be done, so
 				// queue DONT_HAVE and move on
-				provider.taskQueue.PushTasks(sender, peertask.Task{
-					Topic:    topicDontHave(entry.Cid),
-					Priority: int(entry.Priority),
-					Work:     message.BlockPresenceSize(entry.Cid),
-				})
+				provider.queueDontHave(ctx, sender, entry, "failed_retriever_request")
 				continue
 			}
 
 			provider.blockManager.GetAwait(ctx, entry.Cid, func(block blocks.Block) {
-				stats.Record(context.Background(), metrics.BitswapResponseCount.M(1))
-				provider.taskQueue.PushTasks(sender, peertask.Task{
-					Topic:    topicBlock(block),
-					Priority: int(entry.Priority),
-					Work:     len(block.RawData()),
-				})
+				provider.queueBlock(ctx, sender, entry, block)
```

Comment:
`ctx` gets passed into a a callback here, which is stored and invoked later. By the time that happens, `ctx` could already be closed, so for that reason I'd pass `context.Background()` here. But I'm not sure if that messes up the tag for stats recording in that function?

What do you think?

Reply:
Ah, okay. Let me try this again with that in mind. My gut says using `context.Background()` wouldn't change anything. Thanks for the heads up.